src/plugins/intel_cpu/src/nodes/mkldnn_non_zero.cpp

```diff
@@ -108,31 +118,77 @@ void MKLDNNNonZeroNode::executeSpecified() {
-    size_t nonZeroCount = getNonZeroElementsCount(src, inShape);
+    std::vector<size_t> nonZeroCounts = getNonZeroElementsCount(src, inShape);
+    size_t totalNonZeroCount = 0;
+    for (const auto& nonZeroCount : nonZeroCounts)
+        totalNonZeroCount += nonZeroCount;
 
     if (isDynamicNode()) {
-        VectorDims newDims{inRank, nonZeroCount};
+        VectorDims newDims{inRank, totalNonZeroCount};
         redefineOutputMemory({newDims});
     }
     int *dst = reinterpret_cast<int *>(dstMemPtr->GetPtr());
     size_t inSize = inShape.getElementsCount();
-    auto srcStrides = getParentEdgeAt(0)->getMemory().GetDescWithType<BlockedMemoryDesc>()->getStrides();
-    if (nonZeroCount == 0)
+    auto srcDims = inShape.getDims();
+    if (totalNonZeroCount == 0)
         return;
-    if (inShape.getRank() == 0) {
+
+    if (inRank == 0) {
         dst[0] = 0;
     } else {
-        size_t colIndex = 0, outIndex = 0;
-        for (size_t i = 0; i < inSize; i++) {
-            if (src[i] != zero) {
-                size_t temp = i;
-                for (size_t j = 0; j < inRank; j++) {
-                    outIndex = j * nonZeroCount + colIndex;
-                    dst[outIndex] = temp / srcStrides[j];
-                    temp = temp % srcStrides[j];
+        // TODO: find a proper place for this
+        constexpr static const std::size_t MaxDimensions = 6;
+
+        int nthr = static_cast<int>(nonZeroCounts.size());
+        std::vector<size_t> destIndices;
+
+        size_t nonZeroSum = 0;
+        destIndices.resize(nonZeroCounts.size());
+        for (size_t i = 0; i < nonZeroCounts.size(); ++i) {
+            destIndices[i] = nonZeroSum;
+            nonZeroSum += nonZeroCounts[i];
+        }
+
+        if (inRank == 1) {
+            dnnl::impl::parallel(nthr, [&](int ithr, int nthr) {
+                size_t colIndex = destIndices[ithr];
+                dnnl::impl::for_nd(ithr, nthr, inSize, [&](size_t i) {
+                    if (src[i] != zero) {
+                        dst[colIndex] = static_cast<int>(i);
+                        colIndex++;
+                    }
+                });
+            });
+        } else if (inRank <= MaxDimensions) {
+            dnnl::impl::parallel(nthr, [&](int ithr, int nthr) {
+                size_t start = 0, end = 0;
+                dnnl::impl::balance211(inSize, nthr, ithr, start, end);
+
+                size_t startForIndices = start;
+                size_t indices[MaxDimensions];
+                for (int i = static_cast<int>(inRank - 1); i >= 0; i--) {
+                    indices[i] = startForIndices % srcDims[i];
+                    startForIndices = startForIndices / srcDims[i];
+                }
+                indices[inRank - 1] -= 1;
+
+                size_t colIndex = destIndices[ithr], outIndex = 0;
+                for (size_t i = start; i < end; i++) {
+                    for (int j = static_cast<int>(inRank - 1); j >= 0 && ++indices[j] >= srcDims[j]; j--) {
```

Comment:
I don't think indices are properly calculated here.
One of the approaches which should work is to calculate an array of destination indices and iterate over n-dim array in nested loops:
```C++
for (int batch = startIndices[0]; batch < endIndices[0]; batch++)
    for (int channel = startIndices[1]; channel < endIndices[1]; channel++)
        for (int y = startIndices[2]; y < endIndices[2]; y++)
            for (int x = startIndices[2]; x < endIndices[2]; x++) {
                auto inputIndex = batch * batchStride + channel * channelStride + y * yStride + x;
                if (input[inputIndex])
                {
                    output[outputIndedx] = batch;
                    output[batchOffset + outputIndedx] = channel;
                    output[yOffset + outputIndedx] = y;
                    output[xOffset + outputIndedx] = x;
                    outputIndedx++;
                }
            }
             
```